webui/react/src/stores/workspaces.tsx

```diff
@@ -1,69 +1,72 @@
 import { V1PostWorkspaceRequest } from 'services/api-ts-sdk';
 import { GetWorkspacesParams } from 'services/types';
+import { isEqual } from 'shared/utils/data';
 import { Workspace } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
 
-type WorkspacesContext = {
-  updateWorkspaces: (fn: (ws: Loadable<Workspace[]>) => Loadable<Workspace[]>) => void;
-  workspaces: Loadable<Workspace[]>;
-};
+type WorkspacesContext = { workspaces: WritableObservable<Loadable<Workspace[]>> };
 
 const WorkspacesContext = createContext<WorkspacesContext | null>(null);
 
 export const WorkspacesProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [workspaces, updateWorkspaces] = useState<Loadable<Workspace[]>>(NotLoaded);
-  return (
-    <WorkspacesContext.Provider value={{ updateWorkspaces, workspaces }}>
-      {children}
-    </WorkspacesContext.Provider>
-  );
+  const workspaces: WritableObservable<Loadable<Workspace[]>> = observable(NotLoaded);
+
+  return <WorkspacesContext.Provider value={{ workspaces }}>{children}</WorkspacesContext.Provider>;
 };
 
 export const useFetchWorkspaces = (canceler: AbortController): (() => Promise<void>) => {
   const context = useContext(WorkspacesContext);
+
   if (context === null) {
     throw new Error('Attempted to use useFetchWorkspaces outside of Workspace Context');
   }
-  const { updateWorkspaces } = context;
 
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getWorkspaces({}, { signal: canceler.signal });
-      updateWorkspaces(() => Loaded(response.workspaces));
-    } catch (e) {
-      handleError(e);
-    }
-  }, [canceler, updateWorkspaces]);
-};
-
-export const useEnsureWorkspacesFetched = (canceler: AbortController): (() => Promise<void>) => {
-  const context = useContext(WorkspacesContext);
-  if (context === null) {
-    throw new Error('Attempted to use useEnsureWorkspacesFetched outside of Workspace Context');
-  }
-  const { workspaces, updateWorkspaces } = context;
+  const { workspaces } = context;
 
   return useCallback(async (): Promise<void> => {
-    if (workspaces !== NotLoaded) return;
     try {
       const response = await getWorkspaces({}, { signal: canceler.signal });
-      updateWorkspaces(() => Loaded(response.workspaces));
+
+      workspaces.set(Loaded(response.workspaces));
     } catch (e) {
       handleError(e);
     }
-  }, [canceler, workspaces, updateWorkspaces]);
+  }, [canceler, workspaces]);
 };
 
 export const useWorkspaces = (params?: GetWorkspacesParams): Loadable<Workspace[]> => {
   const context = useContext(WorkspacesContext);
+
   if (context === null) {
     throw new Error('Attempted to use useWorkspaces outside of Workspace Context');
   }
-  return Loadable.map(context.workspaces, (workspaces: Workspace[]) =>
+
+  const { workspaces } = context;
+  const [workspacesState, setWorkspacesState] = useState(workspaces.get());
```

Comment:
If you want to memoize the value, [there is a function for that](https://github.com/determined-ai/determined/blob/master/webui/react/src/utils/observable.ts#L20). The version you have here has a subtle bug.

You actually wouldn't need any memoization though with a correctly written `ensureFetched` function.

Reply:
A correctly written`ensureFetched` will make sure the observable is only updated _once_ for data-fetching and so the memoization will be unnecessary since the data isn't changing.